### PR TITLE
Added custom exceptions.

### PR DESCRIPTION
**Pull request by Srujan-D**
1) Codes for the exceptions are in (gennav.custom_exceptions.utils)
2) Exceptions added for:
    1. Key_Error- custom messege will be generated when key is not found in the dictionary containing all the parameters.
    2. Start state is in obstacle.
    3. End/Goal state is in obstacle.
    4. Path is not found by the planner.
    5. Sampler is unable to sample a state.
3) These exceptions have been integrated with all the path planners.
4) Generated docs have been added.

### CODE REVIEW

**Comment by Srujan-D** [comment]
1) In custom exceptions file, try/except should work for py2 and py3 right? Its still throwing a syntax error (see travis ci logs for py2). 
2) I don't know why i'm getting the lint error for incorrect sorting in polygon_env. It isn't showing any message when I do it locally.
4) I have added Key_Error custom exception right now. But need to discuss before using it (I'll need to make appropriate changes in all the plan methods to integrate it).
3) @sushant1212  I think there's some bug in informed_rrtstar.py. It's returning a single state path sometimes.

**Comment by sushant1212** [comment]
>     1. In custom exceptions file, try/except should work for py2 and py3 right? Its still throwing a syntax error (see travis ci logs for py2).
> 
>     2. I don't know why i'm getting the lint error for incorrect sorting in polygon_env. It isn't showing any message when I do it locally.
> 
>     3. I have added Key_Error custom exception right now. But need to discuss before using it (I'll need to make appropriate changes in all the plan methods to integrate it).
> 
>     4. @sushant1212  I think there's some bug in informed_rrtstar.py. It's returning a single state path sometimes.

Whenever solution is not found, then it returns only the start point. That's why you're getting single state paths. Atharv told that we would replace it later with PathNotFound error

**Comment by Srujan-D** [comment]
> Whenever solution is not found, then it returns only the start point. That's why you're getting single state paths. Atharv told that we would replace it later with PathNotFound error

But even when this error is not raised, the path visualized is similar to this:
Screenshot from 2020-08-02 22-39-49

**Comment by sushant1212** [comment]
> > Whenever solution is not found, then it returns only the start point. That's why you're getting single state paths. Atharv told that we would replace it later with PathNotFound error
> 
> But even when this error is not raised, the path visualized is similar to this:
> Screenshot from 2020-08-02 22-39-49

Ok I'll look into it

**Comment by threewisemonkeys-as** [comment]
1. You are using f-strings which arent supported by y2
2. Is tox running ok locally for lint?
3. No need for the key error as we discussed, it can be removed

**Comment by Srujan-D** [comment]
>     1. You are using f-strings which arent supported by y2

>     2. Is tox running ok locally for lint?
              
>     3. No need for the key error as we discussed, it can be removed
              
1. Yeah, but that's why I'm doing `try:` / `except SyntaxError:` . I thought that will work, but it isn't. I can ofc just have `return "%s"%(var)`, but idk if we should have that for py3 (doesn't give an error, but its old).
2. Nvm, lint errors are sorted.
3. Yeah, I'll remove that.

**Comment by threewisemonkeys-as** [comment]
> Yeah, but that's why I'm doing try: / except SyntaxError: . I thought that will work, but it isn't. I can ofc just have return "%s"%(var), but idk if we should have that for py3 (doesn't give an error, but its old). 

Stick to the old format. We should avoid using try accept for python versions wherever possible

File: gennav/utils/custom_exceptions.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
+class Error(Exception):
```
Comment (threewisemonkeys-as) [inline gennav/utils/custom_exceptions.py]: Is this neccessary, cant they all inherit `Exception`  directly

File: gennav/utils/custom_exceptions.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        self.dict = kwargs
+
+    def __str__(self):
+        try:
+            return "%s key not found in the dictionary containing all parameters (dict_keys[%s])" % (self.key, list(self.dict.keys()))
+        except SyntaxError:
+            return f"{self.key} key not found in the dictionary containing all parameters ({self.dict.keys()})."
+
+
+class StartStateinObs(Error):
```
Comment (threewisemonkeys-as) [inline gennav/utils/custom_exceptions.py]: could you name this `InvalidStateState`

File: gennav/utils/custom_exceptions.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        super(StartStateinObs, self).__init__()
+
+    def __str__(self):
+        try:
+            return "%s -> %s" % (self.start, self.message)
+        except SyntaxError:
+            return f"{self.start} -> {self.message}"
+
+
+class GoalStateinObs(Error):
```
Comment (threewisemonkeys-as) [inline gennav/utils/custom_exceptions.py]: Similar name change here to `InvalidGoalState`

**Review by threewisemonkeys-as — CHANGES_REQUESTED** [review]

File: gennav/utils/custom_exceptions.py (hunk 2 above)
Comment (Srujan-D) [reply gennav/utils/custom_exceptions.py]: Yeah.

File: gennav/utils/custom_exceptions.py (hunk 3 above)
Comment (Srujan-D) [reply gennav/utils/custom_exceptions.py]: Yeah.

File: gennav/utils/custom_exceptions.py (hunk 1 above)
Comment (Srujan-D) [reply gennav/utils/custom_exceptions.py]: Not necessary. But in gen, almost in all custom exceptions examples (including the python docs), there's a base Error class similar to this. Should I remove it?

**Comment by Srujan-D** [comment]
Also, getting an `ImportError` in py2 after moving custom exceptions from `.utils` to `.utils.exceptions`.

**Comment by threewisemonkeys-as** [comment]
Try to use absolute imports instead of relative ones

File: gennav/utils/custom_exceptions.py (hunk 1 above)
Comment (threewisemonkeys-as) [reply gennav/utils/custom_exceptions.py]: Yeah, it might be useful if we abstract things into the base class. For now you can remove it

**Comment by Srujan-D** [comment]
I have used absolute imports only, forgot to add gennav.utils.... here. Also, the ImportError is for py2 only. For py3, its working fine.

File: gennav/utils/custom_exceptions.py (hunk 1 above)
Comment (Srujan-D) [reply gennav/utils/custom_exceptions.py]: Okay.

**Comment by threewisemonkeys-as** [comment]
Actuallly, you can keept the excpetions in `utils/exception.py` and add `import gennav.utils.exceptions` in `utils/__init__.py`

**Comment by Srujan-D** [comment]
Cool, I'll do that.

**Comment by Srujan-D** [comment]
> Actuallly, you can keept the excpetions in `utils/exception.py` and add `import gennav.utils.exceptions` in `utils/__init__.py`

It still gives this import error: 
`gennav/utils/__init__.py:2: in <module>`
`import gennav.utils.exceptions.custom_exceptions  # noqa: F401`
`E   ImportError: No module named exceptions.custom_exceptions`

**Comment by Srujan-D** [comment]
> Actuallly, you can keept the excpetions in `utils/exception.py` and add `import gennav.utils.exceptions` in `utils/__init__.py`

Sort, the only fail in the build is for the I-RRT* which sushant resolved in his PR i think.

**Review by threewisemonkeys-as — APPROVED** [review]

**Pull request merged**